### ISSUE TITLE
after gpexpand, dml plan is not correct under orca

### ISSUE THREAD

**Issue by kainwen** [greenplum-db/gpdb#9099]
### Greenplum version or build

GPDB 5X

### Step to reproduce the behavior

1. compile with orca, and build a demo cluster (I test without mirror and standby)
2. create some tables:
```sql
CREATE TABLE sales (id int, date date, amt decimal(10,2))
DISTRIBUTED BY (id)
PARTITION BY RANGE (date)
( START (date '2016-01-29') INCLUSIVE
   END (date '2016-03-03') EXCLUSIVE
   EVERY (INTERVAL '1 day') );

insert into sales values (generate_series(1,100),'2016-02-01',10) ;
insert into sales values (generate_series(1,100),'2016-02-01',10) ;
insert into sales values (generate_series(1,100),'2016-02-01',10) ;
insert into sales values (generate_series(1,100),'2016-03-01',10) ;
create table sales1 as select * from sales ;
```
3. do a gpexpand to enlarge the cluster

```
gpadmin@zlv:~/workspace/test$ cat gpexpand_inputfile_20190123_131719
    zlv:zlv:25435:/home/gpadmin/workspace/test/p/demoDataDir3:5:3:p
gpexpand -i gpexpand_inputfile_20190123_131719 -D gpadmin
```

4. the sql `delete from sales a using sales1 b where a.id=b.id and a.date=b.date and a.amt=b.amt;`
Will throw an error under orca. It works under planner.

```
gpadmin=# set optimizer = off;
SET
gpadmin=# begin;
BEGIN
gpadmin=# delete from sales a using sales1 b where a.id=b.id and a.date=b.date and a.amt=b.amt;
DELETE 400
gpadmin=# abort;
ROLLBACK
gpadmin=# set optimizer = on;
SET
gpadmin=# begin;
BEGIN
gpadmin=# delete from sales a using sales1 b where a.id=b.id and a.date=b.date and a.amt=b.amt;
NOTICE:  Releasing segworker groups to finish aborting the transaction.
ERROR:  could not read block 0 of relation 1663/16384/16388: read only 0 of 32768 bytes  (seg3 127.0.1.1:25435 pid=41620)
ERROR:  could not connect to segment: initialization of segworker group failed (cdbgang.c:238)
```

--------------------------------------

My thoughts:

In GPDB5, for partition tables(originally hash distributed), after phase 1 of gpexpand(adding nodes into cluster), 
root partition's policy is untouched, however, all leaf partition's policy is set randomly.

Delete or update statement joining with other tables in gpdb, we might add motion above result relation. However, dml nodes delete|update using ctid and ctid only works locally (cannot be motioned here).

Orca's plan:
```
gpadmin=# explain delete from sales a using sales1 b where a.id=b.id and a.date=b.date and a.amt=b.amt;
                                                               QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------
 Delete  (cost=0.00..862.09 rows=1 width=1)
   ->  Result  (cost=0.00..862.06 rows=1 width=30)
         ->  Partition Selector for sales  (cost=0.00..862.06 rows=1 width=26)
               ->  Redistribute Motion 4:4  (slice2; segments: 4)  (cost=0.00..862.06 rows=1 width=26)
                     Hash Key: public.sales.id
                     ->  Hash Join  (cost=0.00..862.06 rows=1 width=26)
                           Hash Cond: sales1.id = public.sales.id AND sales1.date = public.sales.date AND sales1.amt = public.sales.amt
                           ->  Table Scan on sales1  (cost=0.00..431.00 rows=100 width=15)
                           ->  Hash  (cost=431.00..431.00 rows=1 width=26)
                                 ->  Broadcast Motion 4:4  (slice1; segments: 4)  (cost=0.00..431.00 rows=1 width=26)
                                       ->  Sequence  (cost=0.00..431.00 rows=1 width=26)
                                             ->  Partition Selector for sales (dynamic scan id: 1)  (cost=10.00..100.00 rows=25 width=4)
                                                   Partitions selected: 34 (out of 34)
                                             ->  Dynamic Table Scan on sales (dynamic scan id: 1)  (cost=0.00..431.00 rows=1 width=26)
 Optimizer status: PQO version 3.83.0
```

We add a broadcast motion above the result relation which might lead to this bug.

---------------------

Contents of gppolicy after gpexpand (1 phase , have not redistribute data)

```
gpadmin=# select localoid::regclass::text, * from gp_distribution_policy ;
        localoid        | localoid | attrnums
------------------------+----------+----------
 sales                  |    16385 | {1}
 gpexpand.status        |    16529 | {1}
 gpexpand.status_detail |    16535 | {1}
 sales1                 |    16525 |
 sales_1_prt_1          |    16388 |
 sales_1_prt_2          |    16393 |
 sales_1_prt_3          |    16397 |
 sales_1_prt_4          |    16401 |
 sales_1_prt_5          |    16405 |
 sales_1_prt_6          |    16409 |
 sales_1_prt_7          |    16413 |
 sales_1_prt_8          |    16417 |
 sales_1_prt_9          |    16421 |
 sales_1_prt_10         |    16425 |
 sales_1_prt_11         |    16429 |
 sales_1_prt_12         |    16433 |
 sales_1_prt_13         |    16437 |
 sales_1_prt_14         |    16441 |
 sales_1_prt_15         |    16445 |
 sales_1_prt_16         |    16449 |
 sales_1_prt_17         |    16453 |
 sales_1_prt_18         |    16457 |
 sales_1_prt_19         |    16461 |
 sales_1_prt_20         |    16465 |
 sales_1_prt_21         |    16469 |
 sales_1_prt_22         |    16473 |
 sales_1_prt_23         |    16477 |
 sales_1_prt_24         |    16481 |
 sales_1_prt_25         |    16485 |
 sales_1_prt_26         |    16489 |
 sales_1_prt_27         |    16493 |
 sales_1_prt_28         |    16497 |
 sales_1_prt_29         |    16501 |
 sales_1_prt_30         |    16505 |
 sales_1_prt_31         |    16509 |
 sales_1_prt_32         |    16513 |
 sales_1_prt_33         |    16517 |
 sales_1_prt_34         |    16521 |
```

**Comment by asimrp** [comment]
Why does `gpexpand` keep the root table's policy unchanged?

**Comment by kainwen** [comment]
> Why does `gpexpand` keep the root table's policy unchanged?

It is a hack so that we can concurrently redistribute leaf tables.

**Comment by zellerh** [comment]
Please see https://github.com/greenplum-db/gporca/pull/555 for a proposed fix for this issue. Please note also that the planner plan will probably generate one segment per partition and that could be a problem for large tables.

**Comment by kainwen** [comment]
> Please see greenplum-db/gporca#555 for a proposed fix for this issue. 

Thanks.

> Please note also that the planner plan will probably generate one slice per partition and that could be a > problem for large tables.

There is another issue to track this.

**Comment by kainwen** [comment]
> Please see [greenplum-db/gporca#555](https://github.com/greenplum-db/gporca/pull/555) for a proposed fix for this issue. Please note also that the planner plan will probably generate one slice per partition and that could be a problem for large tables.

Hi @zellerh , I see the orca's pr is merged. So can we close this?

**Comment by zellerh** [comment]
Yes, I think we can. Thanks for the reminder.
